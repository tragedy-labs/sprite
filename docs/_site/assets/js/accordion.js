/*
 *   This content is licensed according to the W3C Software License at
 *   https://www.w3.org/copyright/software-license-2023/
 *
 *   Simple accordion pattern example
 */

'use strict';

class Accordion {
  constructor(domNode) {
    this.root = domNode;
    this.button = this.root.querySelector('button[aria-expanded]');
    this.icon = this.root.querySelector('svg');
    const controlsId = this.button.getAttribute('aria-controls');
    this.content = document.getElementById(controlsId);

    this.open = this.button.getAttribute('aria-expanded') === 'true';

    // add event listeners
    this.button.addEventListener('click', this.onButtonClick.bind(this));
    this.toggle(false);
  }

  onButtonClick() {
    this.toggle(!this.open);
  }

  toggle(open) {
    // don't do anything if the open state doesn't change
    if (open === this.open) {
      return;
    }

    // update the internal state
    this.open = open;

    // handle DOM updates
    this.button.setAttribute('aria-expanded', `${open}`);
    if (open) {
      this.content.removeAttribute('hidden');
      this.icon.classList.add('open');
    } else {
      this.content.setAttribute('hidden', '');
      this.icon.classList.remove('open');
    }
  }

  // Add public open and close methods for convenience
  open() {
    this.toggle(true);
  }

  close() {
    this.toggle(false);
  }
}

const initializeAccordions = () => {
  // init accordions
  const accordions = document.querySelectorAll('.menuItem');
  accordions.forEach((accordionEl) => {
    new Accordion(accordionEl);
  });
};

document.addEventListener('DOMContentLoaded', initializeAccordions);
