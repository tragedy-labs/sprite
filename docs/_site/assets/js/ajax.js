// Constants
const CONTENT_CONTAINER_ID = "content";
const PARSER = new DOMParser();

// State
let historyStack = [];
let currentStateIndex = -1;

// Functions
async function fetchDocument(url) {
  try {
    const response = await fetch(url);
    return await response.text();
  } catch (error) {
    throw new Error(`Failed to fetch document: ${url}`, { cause: error });
  }
}

async function parseHtml(html) {
  return PARSER.parseFromString(html, "text/html");
}

async function updateContent(htmlDoc) {
  const title = htmlDoc.querySelector("title").textContent;
  const contentDiv = htmlDoc.getElementById(CONTENT_CONTAINER_ID);
  const target = document.getElementById(CONTENT_CONTAINER_ID);
  target.innerHTML = contentDiv.innerHTML;
  target.scrollTop = 0;
  document.title = title;

  // Reinitialize link click events for newly loaded content
  createLinkClickEvents(target);
}

function pushToHistory(url) {
  historyStack = historyStack.slice(0, currentStateIndex + 1);
  historyStack.push(url);
  currentStateIndex++;
  history.pushState({ index: currentStateIndex }, "", url);
}

async function navigate(url, addToHistory = true) {
  const urlObj = new URL(url, window.location.origin);

  if (urlObj)
    if (urlObj.hash && urlObj.pathname === window.location.pathname) {
      // Handle local page anchor navigation
      const anchor = document.querySelector(urlObj.hash);
      if (anchor) {
        anchor.scrollIntoView();
        addToHistory && pushToHistory(url);
        return;
      }
    }

  // Handle Local Navigation
  const html = await fetchDocument(url);
  const htmlDoc = await parseHtml(html);
  await updateContent(htmlDoc);

  addToHistory && pushToHistory(url);
}

function createLinkClickEvents(root) {
  const links = root.querySelectorAll("a");
  links.forEach((link) => {
    link.addEventListener("click", (e) => {
      const url = e.currentTarget.href;
      if (isLocalLink(url)) {
        e.preventDefault();
        navigate(url);
      }
    });
  });
}

function isLocalLink(url) {
  const linkUrl = new URL(url, window.location.origin);
  return linkUrl.origin === window.location.origin;
}

window.addEventListener("popstate", (e) => {
  if (e.state && e.state.index !== undefined) {
    currentStateIndex = e.state.index;
    navigate(historyStack[currentStateIndex], false);
  }
});

// Initialize
function initAjax() {
  createLinkClickEvents(document);
  // Initialize the first state
  historyStack.push(window.location.href);
  currentStateIndex++;
  history.replaceState({ index: currentStateIndex }, "", window.location.href);
}

// Call init
document.addEventListener("DOMContentLoaded", initAjax);
