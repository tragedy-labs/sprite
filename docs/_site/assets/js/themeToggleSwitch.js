function initializeThemeToggle() {
  const element = document.getElementById("theme_toggle");
  const theme = localStorage.getItem("theme");

  const prefersTheme = (theme) =>
    window.matchMedia(`(prefers-color-scheme: ${theme})`).matches;

  const setTheme = (theme) => {
    localStorage.setItem("theme", theme);
    document.body.className = theme;
    setSwitchState(theme);
  };

  const setSwitchState = (theme) => {
    switch (theme) {
      case "dark":
        element.checked = true;
        break;
      case "light":
        element.checked = false;
        break;
      default:
        throw new Error("unknown theme");
    }
  };

  const determinePosition = () => {
    if (theme) {
      setTheme(theme);
    } else {
      if (prefersTheme("dark")) {
        setTheme("dark");
        return;
      } else {
        setTheme("light");
      }
    }
  };

  element.addEventListener("click", () => {
    if (element.checked) {
      setTheme("dark");
    } else {
      setTheme("light");
    }
  });

  determinePosition();
}

document.addEventListener("DOMContentLoaded", initializeThemeToggle);
