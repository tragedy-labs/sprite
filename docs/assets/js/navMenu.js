document.addEventListener("DOMContentLoaded", function () {
  const content = document.getElementById("content");
  const mainNav = document.getElementById("main_nav");
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");
  const links = mainNav.querySelectorAll("a");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let currentTranslation = 0;
  let previousTouchPosition = null;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation < -80) {
      closeMenu();
    }
    currentTranslation = 0;
    mainNav.style.transform = null;
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;

    const mainNavWidth = mainNav.offsetWidth;
    const ratio = (mainNavWidth + currentTranslation) / mainNavWidth;

    content.style.opacity = 0.65 * (1 - ratio) + 0.35;
  }

  function escHandler(e) {
    if (e.key === "Escape") {
      closeMenu();
    }
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
    content.addEventListener("touchstart", closeMenu);
    content.addEventListener("mousedown", closeMenu);
    document.addEventListener("keydown", escHandler);
    // TODO: This could be handled in the ajax script
    // since it already handles link clicks
    links.forEach((link) => {
      link.addEventListener("click", closeMenu);
    });
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
    content.removeEventListener("touchstart", closeMenu);
    content.removeEventListener("mousedown", closeMenu);
    document.removeEventListener("keydown", escHandler);
    links.forEach((link) => {
      link.removeEventListener("click", closeMenu);
    });
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
    content.style.opacity = null;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    content.classList.toggle("fade");
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
