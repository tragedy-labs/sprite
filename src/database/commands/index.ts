export { EXPLAIN } from './EXPLAIN.js';
export { SELECT_SCHEMA } from './SELECT_SCHEMA.js';
