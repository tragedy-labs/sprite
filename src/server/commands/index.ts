export { CLOSE_DATABASE } from './CLOSE_DATABASE.js';
export { CONNECT_CLUSTER } from './CONNECT_CLUSTER.js';
export { CREATE_DATABASE } from './CREATE_DATABASE.js';
export { CREATE_USER, type IArcadeCreateUser } from './CREATE_USER.js';
export { DISCONNECT_CLUSTER } from './DISCONNECT_CLUSTER.js';
export { DROP_DATABASE } from './DROP_DATABASE.js';
export { DROP_USER } from './DROP_USER.js';
export { GET_SERVER_EVENTS } from './GET_SERVER_EVENTS.js';
export { OPEN_DATABASE } from './OPEN_DATABASE.js';
export { SHUTDOWN } from './SHUTDOWN.js';
