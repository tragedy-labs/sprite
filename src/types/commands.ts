import {
  ArcadeRecordCategory,
  EdgeRecordMeta,
  RecordMeta
} from './database.js';
import { DeleteFromCount } from './operators.js';

/** Base Interface for a command response from ArcadeDB */
interface CommandResponse<T = string> {
  operation: T;
}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a command that is a `BUCKET` operation
 */
type BucketOperationResponse<T = string, R = string> = Array<
  CommandResponse<R> & {
    bucketName: T;
    bucketId: number;
  }
>;

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `ALTER BUCKET` command
 */
export interface AlterBucket<Name = string>
  extends BucketOperationResponse<Name, 'alter bucket'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `CREATE BUCKET` command
 */
export interface CreateBucket<Name = string>
  extends BucketOperationResponse<Name, 'create bucket'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `DROP BUCKET` command
 */
export interface DropBucket<Name = string>
  extends BucketOperationResponse<Name, 'drop bucket'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a command that is a `TYPE` operation
 */
type TypeOperationResponse<T = string, R = string> = Array<
  CommandResponse<R> & {
    typeName: T;
  }
>;

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `ALTER TYPE` command
 */
export interface AlterType<Name = string>
  extends TypeOperationResponse<Name, 'alter type'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `CREATE document TYPE <Name>` command
 */
export interface CreateDocumentType<Name = string>
  extends TypeOperationResponse<Name, 'create document type'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `CREATE edge TYPE <Name>` command
 */
export interface CreateEdgeType<Name = string>
  extends TypeOperationResponse<Name, 'create edge type'> {}

/**
 * An array, which contain an object returned from
 * ArcadeDB for a `CREATE vertex TYPE <Name>` command
 */
export interface CreateVertexType<Name = string>
  extends TypeOperationResponse<Name, 'create vertex type'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `DROP document TYPE <Name>` command
 */
export interface DropType<Name = string>
  extends TypeOperationResponse<Name, 'drop type'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `INSERT INTO <document-type>` or
 * a `INSERT INTO <vertex-type>` command
 */
type InsertRecord<T, C extends ArcadeRecordCategory> = Array<T & RecordMeta<C>>;

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `INSERT INTO <document-type>`
 * command
 * @template Type The type of the document to be inserted
 */
export interface InsertDocument<Type = unknown>
  extends InsertRecord<Type, 'd'> {}
/**
 * An array, which contains an object returned from
 * ArcadeDB for a `INSERT INTO <vertex-type>`
 * command
 * @template Type The type of the vertex to be inserted
 */
export interface InsertVertex<Type = unknown> extends InsertRecord<Type, 'v'> {}

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `CREATE EDGE <edge-type>` command
 * @template TypeName - The name of the edge type
 */
export type CreateEdge<Type = unknown> = Array<Type & EdgeRecordMeta>;

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `CREATE VERTEX <vertex-type>` command
 * @template Type - The type of the vertex to create
 */
export type CreateVertex<Type = unknown> = InsertVertex<Type>;

/**
 * An array, which contains an object returned from
 * ArcadeDB for a `DELETE FROM` command. The object
 * contains one property, `count` which is an
 * integer
 */
export type DeleteFrom = Array<DeleteFromCount>;
