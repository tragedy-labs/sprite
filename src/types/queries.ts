import { ArcadeRecordCategory, RecordMeta } from './database.js';

/**
 * The `DOCUMENT`, `EDGE`, or `VERTEX` with ArcadeDB
 * meta attached. If being used in a query command,
 * expect it to be delivered as an array of the
 * described object.
 */
type ArcadeRecord<T, C extends ArcadeRecordCategory> = T & RecordMeta<C>;

/**
 * The `DOCUMENT` record of a certain type, with ArcadeDB meta
 * attached.
 *
 * @note If being used in a query command, expect it
 * to be delivered as an array of the described object.
 *
 * @template T The base `DOCUMENT` type definition (without meta)
 */
export type ArcadeDocument<T = unknown> = ArcadeRecord<T, 'd'>;

/**
 * The `EDGE` record of a certain type, with ArcadeDB meta
 * attached.
 *
 * @note If being used in a query command, expect it
 * to be delivered as an array of the described object.
 *
 * @template T The base `EDGE` type definition (without meta)
 */
export type ArcadeEdge<T = unknown> = ArcadeRecord<T, 'e'>;

/**
 * The `VERTEX` record of a certain type, with ArcadeDB meta
 * attached.
 *
 * @note If being used in a query command, expect it
 * to be delivered as an array of the described object.
 *
 * @template T The base `VERTEX` type definition (without meta)
 */
export type ArcadeVertex<T = unknown> = ArcadeRecord<T, 'v'>;
