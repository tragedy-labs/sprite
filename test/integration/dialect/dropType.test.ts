import { CreateDocumentType } from 'src/types/commands.js';
import { testClient } from './testclient.js';
import { ArcadeDocument } from 'src/types/queries.js';

const db = testClient.database;
const invalidTypeName = 'DROP_TYPE_INVALID_TYPE';
const typeName = 'DropTypeTestType';

type DropTypeTestType = {
  aProperty: string;
};

interface DocumentTypes {
  [typeName]: DropTypeTestType;
  [invalidTypeName]: object;
}

describe('SqlDialect.dropType()', () => {
  it(`should drop a type`, async () => {
    /* Arrange */
    await db.command<CreateDocumentType<typeof typeName>>(
      'sql',
      `CREATE document TYPE ${typeName}`
    );

    /* Act & Assert */
    await expect(
      db.query<ArcadeDocument<DropTypeTestType>>(
        'sql',
        `SELECT FROM ${typeName}`
      )
    ).resolves.toHaveLength(0);

    await testClient.dropType<DocumentTypes, typeof typeName>(typeName);

    /* Assert */
    await expect(
      db.query<ArcadeDocument<DropTypeTestType>>(
        'sql',
        `SELECT FROM ${typeName}`
      )
    ).rejects.toMatchSnapshot();
  });

  it(`should propagate errors from the database`, async () => {
    /* Arrange, Act & Assert */
    await expect(
      testClient.dropType<DocumentTypes, typeof invalidTypeName>(
        invalidTypeName
      )
    ).rejects.toMatchSnapshot();
  });
});
