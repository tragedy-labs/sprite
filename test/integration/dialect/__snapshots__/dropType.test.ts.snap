// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SqlDialect.dropType() should drop a type 1`] = `[Error: Internal error. Type with name 'DropTypeTestType' was not found. Status: 500]`;

exports[`SqlDialect.dropType() should propagate errors from the database 1`] = `[Error: Unable to drop type [DROP_TYPE_INVALID_TYPE] from database [SpriteIntegrationTesting]]`;
